src/lib.rs

```diff
@@ -156,26 +132,33 @@ mod tests {
     use crate::*;
 
     // https://github.com/bitcoin/bitcoin/blob/b264410e012a61b103e1a03c43df4e17b9b75452/src/test/util/setup_common.h#L80
-    const CENT:u64 = 1_000_000;
+    const CENT: u64 = 1_000_000;
+
+    const FEE: u64 = 148;
 
     const UTXO_POOL: [MinimalUtxo; 4] = [
-        MinimalUtxo { value: CENT },
-        MinimalUtxo { value: 2 * CENT },
-        MinimalUtxo { value: 3 * CENT },
-        MinimalUtxo { value: 4 * CENT },
+        MinimalUtxo { value: CENT, fee: FEE },
+        MinimalUtxo { value: 2 * CENT, fee: FEE },
+        MinimalUtxo { value: 3 * CENT, fee: FEE },
+        MinimalUtxo { value: 4 * CENT, fee: FEE },
     ];
 
     const COST_OF_CHANGE: u64 = 500_000;
 
     #[derive(Clone, Debug, Eq, PartialEq)]
     struct MinimalUtxo {
         value: u64,
+        fee: u64,
     }
 
     impl Utxo for MinimalUtxo {
         fn get_value(&self) -> u64 {
             self.value
         }
+
+        fn get_effective_value(&self) -> u64 {
+            self.value - self.fee
+        }
```

Comment:
@Xekyo I think it's important to note that this is the concrete definition provided in the test suite.  In order to run the tests, a concrete definition must be provided similar to how the code calling this crate would provide the definitions for the abstract UTXO [type](https://github.com/p2pderivatives/rust-bitcoin-coin-selection/blob/405451929568422f7df809e35d6ad8f36fccce90/src/lib.rs#L27).  In other words, since the details of how the UTXO is computed don't specifically concern coin_selection, then the definitions should be provided by the calling crate.  The crate that depends on this crate would provide the method body for `get_effective_value`.  Does that makes sense or do you think the abstract type should change?

Reply:
I would orient myself by how rust-bitcoin defines its UTXO objects given that your crate is aiming to be called from there eventually. If the UTXOs in rust-bitcoin already track _effective\_value_ that would be great of course, but a call to a coin selection function should generally come from a context where a feerate is known and an expected input weight might be more accessible than you think.

Reply:
> I would orient myself by how rust-bitcoin defines its UTXO objects given that your crate is aiming to be called from there eventually.

AFAIK/CT there is no UTXO object defined in rust-bitcoin and I don't think the goal is for rust-bitcoin to call this crate but rather for a user to be able to use it next to rust-bitcoin. 

Also the reason for defining a trait instead of a struct was mainly to avoid requiring a conversion when calling the coin selection or forcing users to have to use the defined struct in their code (which in many cases could be lacking some information that they need). I don't have a strong opinion on which way it should go but thought relevant to give the reasons for the design.

Reply:
I see, thanks for the clarification. Let me try to rephrase the point I was attempting to make. The _effective value_ of UTXOs depends on the feerate that a transaction is being built at. This means that the value can only be calculated dynamically either by the caller before calling the coin selection function, or by the coin selection function itself. Since it's a bit of a specific thing that only coin selection needs, it seems to me that it could be handled on this side of the call.

So really, the caller must pass to coin selection one of three packages of information so it can make an informed decision. Coin selection needs either the 1) _effective\_value_ on UTXOs, 2) the intended _feerate_ plus _weight of each UTXO_, or 3) the intended _feerate_ plus the _type of UTXO_ plus knowledge of the _weight of standard input types_.

Reply:
> Also the reason for defining a trait instead of a struct was mainly to avoid requiring a conversion when calling the coin selection or forcing users to have to use the defined struct in their code (which in many cases could be lacking some information that they need). 

Seems like a good reason to keep the interface abstract.

> This means that the value can only be calculated dynamically either by the caller before calling the coin selection function, or by the coin selection function itself.

Or the other option is it's calculated dynamically by a UTXO library that coin-selection depends on (which is neither the calling application nor the coin selection lib).

> Since it's a bit of a specific thing that only coin selection needs, it seems to me that it could be handled on this side of the call

I think since we need this to move forward and there isn't a UTXO library we can import (and we wouldn't want to require the calling application to know how to calculate this), it would make sense to include the UTXO weight calculation logic in this crate (for now).  However, it seems to me the coinselection algorithm doesn't need to know the specifics of how things like the `effective_value` are computed.  Instead, there are certain properties that must hold true, for example that the `effective_value >= target_value` (please correct me if I'm wrong here).  At any rate, I think a lot of this debate is if a UTXO library would ever be useful outside of coinselection, which would mean moving it to a separate library would be better (in the future).

Reply:
oops I meant `effective_value >= value`.

Reply:
> This means that the value can only be calculated dynamically either by the caller before calling the coin selection function, or by the coin selection function itself..

Personally as a user of the library I'd much prefer the second option, as I don't want to have to start thinking about what effective value is and how I should go to compute it.

>  However, it seems to me the coinselection algorithm doesn't need to know the specifics of how things like the effective_value are computed. Instead, there are certain properties that must hold true, for example that the effective_value >= target_value (please correct me if I'm wrong here). At any rate, I think a lot of this debate is if a UTXO library would ever be useful outside of coinselection, which would mean moving it to a separate library would be better (in the future).

I'm not sure a UTXO library would be useful, IMHO it makes sense to have the computation within this crate. The question then is what information we need the user to provide to be able to do the computation. Of course first would be the fee rate, then we need information about the input weight. I can think of two ways: 

1. an enum per input giving the type of output being spend. For P2PKH or P2WPKH the weight can be computed  easily, for P2SH, P2WSH or P2TR we'd need a parameter like `max_script_length` or `max_witness_len`. We could also add some well known types like 2-of-2 P2WSH (though I'd probably do that later).
2. take an output descriptor (from the [rust-miniscript crate](https://github.com/rust-bitcoin/rust-miniscript)) and just call [max_satisfaction_weight]((https://docs.rs/miniscript/latest/miniscript/descriptor/enum.Descriptor.html#method.max_satisfaction_weight) (and add the other input field weight).

The second option feels cleaner but requires a dependency. Maybe a combination of both with a feature flag for miniscript could be an option, but might be better to just pick one for now (if it'd be me I think I'd got the miniscript way but I don't have a very strong opinion).

Reply:
Thanks for all the feedback, I'm not sure a UTXO lib will be useful either, however for now it seems we all agree it's best to add the needed fee computations in this library.  I'll look at getting this added as a new commit to this PR.

Reply:
This commit is a bit hack and slash right now but I wanted to see what people thing: https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/22/commits/b149195a48234eb5872107fb32b4e48d88670488

I noticed this [commit](https://github.com/rust-bitcoin/rust-bitcoin/commit/6d51e9255b969e142d0fae6d4cf3fcdd8cfae0cc) that was recently added which does a lot of the heavy lifting to compute the input weight.  However this requires the downstream application to use the rust-bitcoin `TxIn`.  Also, using the `TxIn` doesn't seem to be enough to get the value of the input.  The `TxIn` only provides the `Outpoint` which is the transaction index [output](https://github.com/rust-bitcoin/rust-bitcoin/blob/master/bitcoin/src/blockdata/transaction.rs#L52) of the corresponding.  It doesn't seem ideal to require the downstream application to pass the value of the input but that's what I've done for now.  There's probably a better way to structure this but may require a change to the upstream lib.

Reply:
I don't think that will work. IIUC the commit you linked to uses the witness size of the input to compute the weight, but when you are doing coin selection you haven't yet signed the input (since the tx is not built yet) so I don't think it really make sense, unless I'm misunderstanding something. Or you would need to require the user to pre-fill the witness with some fake data but that doesn't seem convenient to me.

Reply:
> the commit you linked to uses the witness size of the input to compute the weight

There are two possible weights for the inputs, either `segwit` or `non-segwit`:

```
pub fn calculate_fee(input: &TxIn, is_segwit: bool, feerate: u64) -> u64 {
   let weight = if is_segwit {
        TxIn::segwit_weight(input)
    } else {
        TxIn::legacy_weight(input)
    };

    weight as u64 * feerate
}
```
 
> but when you are doing coin selection you haven't yet signed the input (since the tx is not built yet) so I don't think it really make sense

I'm not sure about that right now.  My first thought is that I would think the signature is a known size ahead of time depending on if the transaction is segwit or not.  I guess my thinking is that we are only looking to examine the weight of each input to determine the feerate and subtract that feerate from the input value (effective_value).  I'm not sure how signing one or more inputs impacts the calculation.

Reply:
The problem is that you don't know what data is required to unlock the UTXO, it could be a single signature, but it could be several or also some pre-image of a hash or something else. I feel maybe the best for now would be to just have a `get_max_satisfaction_weight` on the UTXO trait to move forward and maybe try to improve later.

Reply:
I see what you mean now.  Ideally we could just call a method on each UTXO and the precise weight would be returned.  For now I guess it can be approximated with `get_max_satisfaction_weight`.

Reply:
Just a quick follow up question to check my understanding.  Is the reason we don't know the data required due to the fact that it's due to the fact that it's part of the witness data which isn't stored?

Reply:
I've been reading a bit more and am still not sure if my understanding is correct.  It seems that a UTXO size isn't known until the party signs and you wouldn't expect the party to provide all possible ways to sign every UTXO.  The follow question assuming this is correct is, why isn't there some type of metadata created when the UTXO is created which can describe the size once it's signed.

Reply:
The output script determines how an input may spend a UTXO. If it's one of the standard scripts, this is sufficient to predict the weight of the input.

- P2PKH: [148 B](https://bitcoin.stackexchange.com/q/48279/5406), 147 B if it has a low-R signature, with a very tiny chance of being even smaller
- P2SH-P2WPKH: [364 WU/91 vB](https://bitcoin.stackexchange.com/q/111392/5406), 363 WU for a low-R signature
- P2WPKH: [272 WU/68 v B](https://bitcoin.stackexchange.com/q/100159/5406), 271 WU for a low-R signature
- P2TR keypath spend: [230 WU/57.5 vB](https://bitcoin.stackexchange.com/q/111395/5406)

Reply:
> The output script determines how an input may spend a UTXO. If it's one of the standard scripts, knowledge of the output script is sufficient to predict the weight of the input.
> 
> * P2PKH: [148 B](https://bitcoin.stackexchange.com/q/48279/5406), 147 B if it has a low-R signature
> * P2SH-P2WPKH: [364 WU/91 vB](https://bitcoin.stackexchange.com/q/111392/5406), 363 WU for a low-R signature
> * P2WPKH: [272 WU/68 v B](https://bitcoin.stackexchange.com/q/100159/5406), 271 WU for a low-R signature
> * P2TR keypath spend: [230 WU/57.5 vB](https://bitcoin.stackexchange.com/q/111395/5406)

Yes so way to take advantage of this would be to go with [option 1](https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/22#discussion_r1068817903) I think.

> Is the reason we don't know the data required due to the fact that it's part of the witness data which isn't stored?

It's because there are many possible scripts that can be used to lock an output which require different inputs of different size.

> The follow question assuming this is correct is, why isn't there some type of metadata created when the UTXO is created which can describe the size once it's signed.

A wallet using a non standard script would probably generate such metadata, but it's not part of the transaction, and I think the closest thing to a standard for this metadata would be an output descriptor.

Reply:
Oh, I did not realize. Do Descriptors specify the input weight of the script they produce?

Reply:
So my understanding from [this](https://bitcoin.stackexchange.com/questions/99540/what-are-output-descriptors#answer-99541) and [this](https://docs.rs/miniscript/latest/miniscript/descriptor/enum.Descriptor.html#method.max_satisfaction_weight) is that you can get an upper bound on the witness size that could be required. Though thinking more now for cases like when you spend a Taproot output through the key path you might be overpaying if you had some script path that you don't use,  but not sure if it's worth considering. If it is then we'd need more information probable.

Reply:
`max_satisfaction_weight` has already been deprecated in favor of `max_weight_to_satisfy` see [here](https://github.com/rust-bitcoin/rust-miniscript/blob/master/src/descriptor/mod.rs#L369).  It sounds like we are OK with just going with descriptor outputs then?  It seems like many wallets are in the process or haven't incorporated descriptors yet.  Also, to answer Xekyo and my previous question, output descriptors store metadata about the derivation path and script key format from my understanding but they don't store metadata about the precise satisfaction size.  Maybe descriptors could be extended to include the satisfaction size of the future so we don't need to use an upper bound estimate?

Reply:
@Tibo-lg: Yeah, the scriptpath would generally be way more expensive. Such behavior may also present a fingerprint that leaks whether a keypath spend has a scriptpath, if people can distinguish what software created the tx.

Reply:
> It sounds like we are OK with just going with descriptor outputs then?
I don't know considering that it's still not enough info to compute the input weight properly I'm not sure it's worth the dependency.

>  Such behavior may also present a fingerprint that leaks whether a keypath spend had a scriptpath alternative
Yeah definitely not ideal.

@yancyribbens maybe just go with something abstract or high level for now and try to get the actual coin selection logic implemented, then we can maybe keep brainstorming on the best way to get the input weight info in an issue or something.

Reply:
@Tibo-lg sure, that makes sense to focus on the selection semantics for now.  Adding weight calculations can come next.

I pushed a new commit that just uses a concrete type for the UTXO and assumes we already know the UTXO weight.  Thoughts?

Reply:
I moved fee_rate into a new struct called Spend in this [commit](https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/22/commits/63b9fbf84621b892a708261e6e5dc4540e4e842f).  This should help determine which params are needed when a UTXO is spent.  Also, I removed `Clone` in this [commit](https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/22/commits/bd45a30131ad9005978bc4da4bac09c9d8bd6170) since there's an overhead to cloning AFAIK.  Every time clone is called it's making a new copy, so we can instead just pass around pointers.

Reply:
>I moved fee_rate into a new struct called Spend in this [commit](https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/22/commits/f18b04f9445b2a8305cc2a8f3aa4aa2572cd2d0a). This should help determine which params are needed when a UTXO is spent.

Are there going to be other value in the `Spend` struct? Otherwise it feels like you're just wrapping a single field which doesn't make a lot of sense to me.

> Also, I removed Clone in this [commit](https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/22/commits/917b6b6ced4a4304e0c3b73aa95bda984616d235) since there's an overhead to cloning AFAIK. Every time clone is called it's making a new copy, so we can instead just pass around pointers.

Sure that's fine!

Reply:
@Tibo-lg ok, maybe Spend is overkill.  How about just a simple fee module? https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/22/commits/4c3137b9a638e78aa09dfead4b8452f351beee29#diff-f4eff9ed318d5e5d7acac95241be35055cf36968693da8a60c16b8b53e97fdd4

I assume we'll want to use the fee helper functions for other algorithms which is why I put it in a separate module.

Reply:
Sure that's fine!